src/operations/find.ts

```diff
@@ -89,169 +92,242 @@ export class FindOperation extends CommandOperation<FindOptions, Document> {
 
+    const serverWireVersion = maxWireVersion(server);
     const options = this.options;
     if (typeof options.allowDiskUse !== 'undefined' && serverWireVersion < 4) {
       callback(new MongoError('The `allowDiskUse` option is not supported on MongoDB < 3.2'));
       return;
     }
 
-    const findCommand: Document = Object.assign({}, this.cmd);
+    if (options.collation && serverWireVersion < SUPPORTS_WRITE_CONCERN_AND_COLLATION) {
+      callback(
+        new MongoError(
+          `Server ${server.name}, which reports wire version ${serverWireVersion}, does not support collation`
+        )
+      );
 
-    if (options.sort) {
-      findCommand.sort = formatSort(options.sort);
+      return;
     }
 
-    if (options.projection) {
-      let projection = options.projection;
-      if (projection && !Buffer.isBuffer(projection) && Array.isArray(projection)) {
-        projection = projection.length
-          ? projection.reduce((result, field) => {
-              result[field] = 1;
-              return result;
-            }, {})
-          : { _id: 1 };
+    if (serverWireVersion < 4) {
+      if (this.readConcern && this.readConcern.level !== 'local') {
+        callback(
+          new MongoError(
+            `server find command does not support a readConcern level of ${this.readConcern.level}`
+          )
+        );
+
+        return;
       }
 
-      findCommand.fields = projection;
+      const findCommand = makeLegacyFindCommand(this.ns, this.filter, options);
+      if (isSharded(server) && this.readPreference) {
+        findCommand.$readPreference = this.readPreference.toJSON();
+      }
+
+      server.query(
+        this.ns,
+        findCommand,
+        {
+          ...this.options,
+          ...this.bsonOptions,
+          documentsReturnedIn: 'firstBatch',
+          readPreference: this.readPreference
+        },
+        callback
+      );
+
+      return;
     }
 
-    if (options.hint) {
-      findCommand.hint = normalizeHintField(options.hint);
+    let findCommand = makeFindCommand(this.ns, this.filter, options);
+    if (this.explain) {
+      findCommand = decorateWithExplain(findCommand, this.explain);
     }
 
-    if (typeof options.skip === 'number') {
-      findCommand.skip = options.skip;
+    server.command(
+      this.ns.toString(),
+      findCommand,
+      {
+        fullResult: !!this.fullResponse,
+        ...this.options,
+        ...this.bsonOptions,
+        documentsReturnedIn: 'firstBatch'
+      },
+      callback
+    );
+  }
+}
+
+function makeFindCommand(ns: MongoDBNamespace, filter: Document, options: FindOptions): Document {
+  const findCommand: Document = {
+    find: ns.collection,
+    filter
+  };
+
+  if (options.sort) {
+    findCommand.sort = formatSort(options.sort);
+  }
+
+  if (options.projection) {
+    let projection = options.projection;
+    if (projection && !Buffer.isBuffer(projection) && Array.isArray(projection)) {
```

Comment:
```suggestion
    if (projection && Array.isArray(projection)) {
```
Nit: Shouldn't `Array.isArray` imply `!Buffer.isBuffer`?

Reply:
fixed